packages/grid/x-data-grid-premium/src/hooks/features/clipboard/useGridClipboardImport.ts

```diff
@@ -0,0 +1,417 @@
+
+    const handlePasteEvent = (event: ClipboardEvent) => {
+      el.removeEventListener('paste', handlePasteEvent);
+      const text = event.clipboardData?.getData('text/plain');
+      if (focusedCell instanceof HTMLElement) {
+        focusedCell.focus({ preventScroll: true });
+      }
+      el.remove();
+      resolve(text || '');
+    };
+
+    el.addEventListener('paste', handlePasteEvent);
+    rootEl.appendChild(el);
+    el.focus({ preventScroll: true });
+  });
+}
+
+const stringToBoolean = (value: string) => {
+  switch (value.toLowerCase().trim()) {
+    case 'true':
+    case 'yes':
+    case '1':
+      return true;
+
+    case 'false':
+    case 'no':
+    case '0':
+    case 'null':
+    case 'undefined':
+      return false;
+
+    default:
+      return undefined;
+  }
+};
+
+const parseCellStringValue = (value: string, colDef: GridColDef) => {
+  switch (colDef.type) {
+    case 'number': {
+      return Number(value);
+    }
+    case 'boolean': {
+      return stringToBoolean(value);
+    }
+    case 'singleSelect': {
+      const colDefValueOptions = (colDef as GridSingleSelectColDef).valueOptions;
+      const valueOptions =
+        typeof colDefValueOptions === 'function'
+          ? colDefValueOptions({ field: colDef.field })
+          : colDefValueOptions || [];
+      const valueOption = valueOptions.find((option) => {
+        if (option === value) {
+          return true;
+        }
+        // TODO: would it work with valueFormatter?
+        if (typeof option === 'object' && option !== null) {
+          if (String(option.label) === value) {
+            return true;
+          }
+        }
+        return false;
+      });
+      if (valueOption) {
+        return valueOption;
+      }
+      return value;
+    }
+    case 'date':
+    case 'dateTime': {
+      const date = new Date(value);
+      return date;
+    }
+    default:
+      return value;
+  }
+};
+
+// Keeps track of updated rows during clipboard paste
+class CellValueUpdater {
+  rowsToUpdate: {
+    [rowId: GridRowId]: GridValidRowModel;
+  } = {};
+
+  updateRow: (row: GridRowModel) => void;
+
+  options: {
+    apiRef: React.MutableRefObject<GridPrivateApiPremium>;
+    processRowUpdate: DataGridPremiumProcessedProps['processRowUpdate'];
+    onProcessRowUpdateError: DataGridPremiumProcessedProps['onProcessRowUpdateError'];
+    getRowId: DataGridPremiumProcessedProps['getRowId'];
+  };
+
+  constructor(options: CellValueUpdater['options']) {
+    this.options = options;
+    this.updateRow = batchRowUpdates(options.apiRef.current.updateRows, 50);
+  }
+
+  updateCell({
+    rowId,
+    field,
+    pastedCellValue,
+  }: {
+    rowId: GridRowId;
+    field: GridColDef['field'];
+    pastedCellValue: string;
+  }) {
+    const { apiRef, getRowId } = this.options;
+    const colDef = apiRef.current.getColumn(field);
+    if (!colDef || !colDef.editable) {
+      return;
+    }
+    if (pastedCellValue === undefined) {
+      return;
+    }
```

Comment:
Nitpick: `pastedCellValue` could possibly be compared in the start to avoid extracting/comparing `colDef` unnecessarily.
```suggestion
    if (pastedCellValue === undefined) {
      return;
    }
    const { apiRef, getRowId } = this.options;
    const colDef = apiRef.current.getColumn(field);
    if (!colDef || !colDef.editable) {
      return;
    }
```